Sources/NIO/Codec.swift

```diff
@@ -96,64 +103,94 @@ private extension ChannelHandlerContext {
 
 extension ByteToMessageDecoder {
 
+    /// Decode in a loop until there is nothing more to decode.
+    private func decodeLoop(ctx: ChannelHandlerContext, decodeFunc: (ChannelHandlerContext, inout ByteBuffer) throws -> DecodingState) throws {
+        while var slice = self.cumulationBuffer?.slice(), slice.readableBytes > 0 {
+            // Needed to later calculate how much we need to advance the readerIndex of the cumulationBuffer.
+            let sliceReadable = slice.readableBytes
+            let sliceWriterIndex = slice.writerIndex
+
+            // We fetch the writerIndex of the cumulationBuffer to make a good guess about if the cumulationBuffer changed in between due re-entrant call to
+            // channelRead after we called decodeFunc(...).
+            let writerIndex = self.cumulationBuffer!.writerIndex
+            let result = try decodeFunc(ctx, &slice)
+
+            guard self.cumulationBuffer != nil else {
+                // The cumulationBuffer was set to nil by either removing the decoder or closing the channel, just break the loop.
+                break
+            }
+
+            guard slice.writerIndex == sliceWriterIndex else {
+                fatalError("Writing to the buffer is not allowed")
```

Comment:
Minor nit, but I tend to prefer these to be `preconditionFailure`.

Reply:
sure why not... I dont care and I think at the end it makes no difference as long as it crashes :)